### PR TITLE
feat: FileInputStream/FileOutputStream instrumentation

### PR DESCRIPTION
**Pull request by romtsn**
_#skip-changelog_

## :scroll: Description

* Introduce new screen in the sample where we utilise File I/O apis
* Introduce new remapping

image

## :green_heart: How did you test it?
Manually for now - automated tests will follow in future PRs

## :pencil: Checklist

- [x] I reviewed the submitted code
- [ ] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

## :crystal_ball: Next steps
The PR is targeting integration branch for now, the next steps are:
* Runtime instrumentation lib with the actual instrumented classes
* More classes instrumentation (FileReader/FileWriter, context.openFileInput/Output)
* Tests

### CODE REVIEW

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/instrumentation/remap/RemappingInstrumentable.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        apiVersion: Int,
+        originalVisitor: ClassVisitor,
+        parameters: SpanAddingClassVisitorFactory.SpanAddingParameters
+    ): ClassVisitor =
+        ClassRemapper(originalVisitor, SimpleRemapper(mapping))
+
+    override fun isInstrumentable(data: ClassContext): Boolean {
+        return when {
+            data.currentClassData.className.startsWith("io.sentry")
+                && !data.currentClassData.className.startsWith("io.sentry.android.roomsample") -> false
```
Comment (romtsn) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/instrumentation/remap/RemappingInstrumentable.kt:34]: this is just for the sample to work, probably shouldn't be in the production code.. maybe worth introducing ignore list config in the gradle plugin, where users can specify packagenames to ignore?

**Comment by marandaneto** [comment]
why `#skip-changelog`? :)

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/instrumentation/remap/RemappingInstrumentable.kt (hunk 1 above)
Comment (marandaneto) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/instrumentation/remap/RemappingInstrumentable.kt:34]: yeah, but I'd rather take it as feature request, not really part of this PR.
we could rename the sample package if necessary to avoid this.
on Java, we special case these packages btw https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/SentryStackTraceFactory.java#L41-L46
because of our samples and mobile app.

**Comment by romtsn** [comment]
> why `#skip-changelog`? :)

targeting integration branch for now

File: examples/android-room/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -16,9 +16,13 @@
             android:name=".ui.EditActivity"
             android:theme="@style/Theme.AppCompat.NoActionBar" />
 
+        <activity
+            android:name=".ui.LyricsActivity"
+            android:theme="@style/Theme.AppCompat.NoActionBar" />
+
         <meta-data
             android:name="io.sentry.dsn"
-            android:value="https://1053864c67cc410aa1ffc9701bd6f93d@o447951.ingest.sentry.io/5428559" />
```
Comment (marandaneto) [inline examples/android-room/src/main/AndroidManifest.xml:21]: why the change of DSN?

File: examples/android-room/src/main/java/io/sentry/android/roomsample/ui/LyricsActivity.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            lyricsInput.setText(file.readText())
+        }
+    }
+
+    override fun onBackPressed() {
+        if (!file.exists()) {
+            file.createNewFile()
+        }
+        file.writeText(lyricsInput.text.toString())
+        transaction.finish(SpanStatus.OK)
```
Comment (marandaneto) [inline examples/android-room/src/main/java/io/sentry/android/roomsample/ui/LyricsActivity.kt:53]: timing of `transaction` wont make any sense here since `onBackPressed` can be clicked hours later or not even clicked, I'd rather use the auto transaction for `LyricsActivity`.
and a new transaction for `onBackPressed` if none set in the scope.

File: examples/android-room/src/main/java/io/sentry/android/roomsample/ui/MainActivity.kt [hunk 4]
```diff
@@ -119,6 +122,16 @@ class TrackAdapter : RecyclerView.Adapter<TrackAdapter.ViewHolder>() {
                 ).putExtra(EditActivity.TRACK_EXTRA_KEY, track)
             )
         }
+        holder.row.infoButton.setOnClickListener {
+            val context = holder.row.context
```
Comment (marandaneto) [inline examples/android-room/src/main/java/io/sentry/android/roomsample/ui/MainActivity.kt:126]: maybe we separate the `TrackAdapter` into a new class, to keep a bit more organized.

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/instrumentation/remap/RemappingInstrumentable.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+package io.sentry.android.gradle.instrumentation.remap
+
+import com.android.build.api.instrumentation.ClassContext
+import io.sentry.android.gradle.instrumentation.ClassInstrumentable
+import io.sentry.android.gradle.instrumentation.SpanAddingClassVisitorFactory
+import org.objectweb.asm.ClassVisitor
+import org.objectweb.asm.commons.ClassRemapper
+import org.objectweb.asm.commons.SimpleRemapper
+
+class RemappingInstrumentable : ClassInstrumentable {
```
Comment (marandaneto) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/instrumentation/remap/RemappingInstrumentable.kt:12]: I like the generic approach

**Comment by romtsn** [comment]
@marandaneto PTAL. We've changed approach so this
```
val fis = FileInputStream("Test.txt")
```

Turns into this
```
val fis = SentryFileInputStream.Factory.create(FileInputStream("Test.txt"), "Test.txt")
```

File: buildSrc/src/main/java/Dependencies.kt [hunk 6]
```diff
@@ -13,7 +13,7 @@ object LibsVersion {
     const val JUNIT = "4.13.2"
     const val ASM = "9.2"
     const val SQLITE = "2.1.0"
-    const val SENTRY = "5.1.2"
+    const val SENTRY = "5.4.0"
```
Comment (marandaneto) [inline buildSrc/src/main/java/Dependencies.kt:16]:
```suggestion
    const val SENTRY = "5.5.0"
```

File: plugin-build/src/test/kotlin/io/sentry/android/gradle/instrumentation/wrap/visitor/WrappingVisitorTest.kt [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    val opcode: Int,
+    val owner: String,
+    val name: String,
+    val descriptor: String,
+    val isInterface: Boolean
+)
+
+data class VarVisit(
+    val opcode: Int,
+    val `var`: Int
```
Comment (marandaneto) [inline plugin-build/src/test/kotlin/io/sentry/android/gradle/instrumentation/wrap/visitor/WrappingVisitorTest.kt:227]: can we use aliases for avoiding backticks?

**Review by marandaneto — APPROVED** [review]
LGTM
